### PR TITLE
Rework routing and URL paths

### PR DESCRIPTION
**Pull request by heralden**
This PR is paired with #362. Basically it changes how routes are handled (as well as library used), adding numerous features/fixes along the way to make sure it was a good choice. [=

Here's a copy of the list of goals and related issues:

- [x] Rename reportpage path
- [x] Query string including keyword for *search* path
- [x] Replace *results* path with *list*
- [x] If a list is not in the history, automatically query for it (such as the case when directly opening path `/humanmine/list/PL_Pax6_Targets`) #66
- [x] Make sure report page deep links work from a cold start in a different mine #78
- [x] Deep linking of mines (try to have the mine namespace optional in the URL path, but fallback to specifying `default` if this isn't practical) #352
- [x] Replace secretary+accountant with reitit for reverse-routing (so we can easily rename routes)
- [x] Use anchor elements and href (to allow *open in new window*) #90
- [x] Use HTML5 history API instead of hash history
- [x] Make sure back button works with the new solution

Something I haven't tested is the creation of lists through saving a query, set operations, etc. Although they should *probably* work as `:results/history+` which they rely on still has the same behaviour.

### CODE REVIEW

**Comment by yochannah** [comment]
@uosl - I'm taking a look at this now, but can I get you to make sure travis passes before we merge?

**Comment by uosl** [comment]
The Cypress test failures are dependent on this fix: https://github.com/metosin/reitit/pull/301

File: src/cljs/bluegenes/core.cljs [hunk 1]
```diff
@@ -26,22 +25,24 @@
...
                   (ocall js/document "getElementById" "app")))
 
-(defn navigate-to-deep-links []
-  (let [url (oget js/window "location" "hash")
-        hashless-path (last (clojure.string/split url #"#"))]
-    (cond (> (count url) 2) ;; if there is more than #/ in the url, navigate there
-          (if (= (first hashless-path) "/")
-            (navigate! hashless-path)
-            (navigate! (str "/" hashless-path))))))
+;; TODO replace or delete this function depending on needs with reitit
```
Comment (yochannah) [inline src/cljs/bluegenes/core.cljs:33]: what are the considerations here with regards to the todo?

File: src/cljs/bluegenes/events.cljs [hunk 2]
```diff
@@ -60,24 +60,52 @@
...
+                                     :name (:name mine-m)
+                                     :id mine-kw}))))
+
+;; This event handler doesn't do anything, as it exists only for `async-flow` to
+;; observe so it knows that `:init-mine` has successfully completed.
+(reg-event-fx
+ :success-init-mine
+ (fn [_ _] {}))
+
+;; This is for setting :current-mine based on the URL path. As opposed to
```
Comment (yochannah) [inline src/cljs/bluegenes/events.cljs:80]: nice comments here!

File: src/cljs/bluegenes/core.cljs (hunk 1 above)
Comment (heralden) [reply src/cljs/bluegenes/core.cljs:33]: Deep linking seems to be working fine without this, so we can delete this code.

Edit: I can see it's specifically for handling links with the hash router, which we don't use anymore.

File: src/cljs/bluegenes/route.cljs [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+  (:require [re-frame.core :refer [subscribe dispatch reg-sub reg-event-db reg-event-fx reg-fx]]
+            [oops.core :refer [ocall]]
+            [reitit.core :as r]
+            [reitit.coercion :as rc]
+            [reitit.coercion.spec :as rcs]
+            [reitit.frontend :as rf]
+            [reitit.frontend.controllers :as rfc]
+            [reitit.frontend.easy :as rfe]))
+
+;; # Quickstart guide:
```
Comment (yochannah) [inline src/cljs/bluegenes/route.cljs:11]: This is fabulous, thank you!

**Comment by yochannah** [comment]
perfect, thanks

File: src/cljs/bluegenes/route.cljs [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+(reg-event-fx
+ ::navigate
+ (fn [{db :db} [_ route & [params query]]]
+   {::navigate! {:k route
+                 :query query
+                 :params (update params :mine #(or % (:current-mine db)))}}))
+
+(reg-event-db
+ ::navigated
```
Comment (yochannah) [inline src/cljs/bluegenes/route.cljs:44]: could we add a comment about when/why this is used?

Comment (heralden) [reply src/cljs/bluegenes/route.cljs:44]: Is it worth it? It's pretty internal to how the router works, so it's not that important to know how it works, and those who do want to know can read it as it's still a pretty short and simple function. [=
I at least can't think of a succinct and useful comment to put there.

**Comment by yochannah** [comment]
"you probably won't need to use this as it's an internal helper method" ?

File: src/cljs/bluegenes/route.cljs (hunk 4 above)
Comment (heralden) [reply src/cljs/bluegenes/route.cljs:44]: >"you probably won't need to use this as it's an internal helper method" ?

Sure, marking it as internal sounds reasonable.

File: src/cljs/bluegenes/route.cljs [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+      [{:start #(dispatch [:set-active-panel :mymine-panel])}]}]
+    ["/report/:type/:id"
+     {:name ::report
+      :controllers
+      [{:parameters {:path [:mine :type :id]}
+        :start (fn [{{:keys [mine type id]} :path}]
+                 (dispatch [:set-active-panel :reportpage-panel
+                            {:type type :id id :mine mine}
+                            [:load-report mine type id]]))}]}]]])
+;; You can do initialisations by adding a :start function to :controllers.
```
Comment (yochannah) [inline src/cljs/bluegenes/route.cljs:203]: I'm not quite sure what you mean here by teardowns and initialisations here, or why I might want to use them.

**Review by yochannah — APPROVED** [review]
:medal_sports:

File: src/cljs/bluegenes/route.cljs (hunk 5 above)
Comment (heralden) [reply src/cljs/bluegenes/route.cljs:203]: Most of the routes have a `:controllers` vector with a `:start` map. While you can copy-cat the existing `:start` maps when you create new routes, there aren't any using `:stop` (teardowns) currently. I wanted to make sure to write it somewhere so people know they exist as well.

**Pull request merged**